### PR TITLE
IDCOM-2102 Controller support

### PR DESCRIPTION
**Pull request by dankelleher**
Cryptid v2 now finally supports controllers. Once we have client docs I will add this there, but example usage is:

Given did:sol:alice, a controller of did:sol:bob:

```
const cryptid = await Cryptid.buildFromDID(
        "did:sol:bob",
        aliceWallet,
        {  connection: provider.connection }
      ).controlWith("did:sol:alice");
```
This also works with controller chaining, so if did:sol:carol is a controller of did:sol:alice, you can do:

```
const cryptid = await Cryptid.buildFromDID(
        "did:sol:bob",
        carolWallet,
        {  connection: provider.connection }
      )
   .controlWith("did:sol:alice")
   .controlWith("did:sol:carol")
```

It is also possible to create a cryptid account for a DID as a controller:
```
Builder.createFromDID(controlledDid, controllerAuthority, { ..., controllerChain: [controllerDid]})
```
This API is different to the .controlWith() API for building cryptid clients, and therefore could do with being rethought a little perhaps.

### CODE REVIEW

File: programs/cryptid/src/instructions/create.rs [hunk 1]
```diff
@@ -46,7 +49,13 @@ pub fn create(
     // convert the controller chain (an array of account indices) into an array of accounts
     // note - cryptid does not need to check that the chain is valid, or even that they are DIDs
     // sol_did does that
-    let controlling_did_accounts = ctx.remaining_accounts.iter().collect::<Vec<&AccountInfo>>();
+    // let controlling_did_accounts = ctx.remaining_accounts.iter().collect::<Vec<&AccountInfo>>();
```
Comment (JulianSpring) [inline programs/cryptid/src/instructions/create.rs]: Can we remove this?

File: programs/cryptid/src/instructions/direct_execute.rs [hunk 2]
```diff
@@ -83,7 +81,19 @@ pub fn direct_execute<'a, 'b, 'c, 'info>(
...
+    let controlling_did_accounts = controller_chain
+        .iter()
+        .map(|controller_reference| {
+            (
+                all_accounts[controller_reference.account_index as usize],
+                controller_reference.authority_key,
+            )
+        })
+        .collect::<Vec<(&AccountInfo, Pubkey)>>();
+    // .as_slice();
```
Comment (JulianSpring) [inline programs/cryptid/src/instructions/direct_execute.rs]: Can we remove this?

**Review by JulianSpring — APPROVED** [review]

File: programs/cryptid/src/instructions/create.rs (hunk 1 above)
Comment (dankelleher) [reply programs/cryptid/src/instructions/create.rs]: Done - thanks

File: programs/cryptid/src/instructions/direct_execute.rs (hunk 2 above)
Comment (dankelleher) [reply programs/cryptid/src/instructions/direct_execute.rs]: Done - thanks

File: packages/client/core/src/api/simpleCryptidClient.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ export class SimpleCryptidClient extends AbstractCryptidClient {
     return this._wallet;
   }
 
-  as(controlledDid: string): ControlledCryptidClient {
-    return new ControlledCryptidClient(controlledDid, this, this.options);
+  controlWith(controllerDid: string): CryptidClient {
```
Comment (rado0x54) [inline packages/client/core/src/api/simpleCryptidClient.ts:20]: actually found the `as` quite elegant, but understand that this is more descriptive.

File: packages/client/core/src/lib/CryptidTransaction.ts [hunk 4]
```diff
@@ -117,15 +136,75 @@ export class CryptidTransaction {
...
       cryptidAccount,
       authority,
       instructions,
-      filteredAccountMetas
+      allAccountMetas,
+      controllerIndices
     );
   }
 
   // TODO move transactionAccountAddress into constructor?
```
Comment (rado0x54) [inline packages/client/core/src/lib/CryptidTransaction.ts]: This TODO was separated from its previous function.

**Review by rado0x54 — APPROVED** [review]

**Comment by rado0x54** [comment]
As mentioned in the meeting, there is come TODOs left around propose execute tests still having `Buffer.from([]), // TODO, support controller chain,`

**Review by rado0x54 — APPROVED** [review]

**Pull request merged**